sympy/codegen/ast.py

```diff
@@ -0,0 +1,391 @@
+     |                                    |--->SubAugmentedAssignment
+     |                                    |--->MulAugmentedAssignment
+     |                                    |--->DivAugmentedAssignment
+     |                                    |--->ModAugmentedAssignment
+     |
+     |--->DataType
+     |           |--------|--->NativeBool
+     |                    |--->NativeInteger
+     |                    |--->NativeFloat
+     |                    |--->NativeDouble
+     |                    |--->NativeVoid
+     |
+     |--->For
+"""
+
+from __future__ import print_function, division
+
+
+from sympy.core import Symbol, Tuple
+from sympy.core.basic import Basic
+from sympy.core.sympify import _sympify
+from sympy.core.relational import Relational
+from sympy.utilities.iterables import iterable
+
+class Assignment(Relational):
+    """
+    Represents variable assignment for code generation.
+
+    Parameters
+    ----------
+    lhs : Expr
+        Sympy object representing the lhs of the expression. These should be
+        singular objects, such as one would use in writing code. Notable types
+        include Symbol, MatrixSymbol, MatrixElement, and Indexed. Types that
+        subclass these types are also supported.
+
+    rhs : Expr
+        Sympy object representing the rhs of the expression. This can be any
+        type, provided its shape corresponds to that of the lhs. For example,
+        a Matrix type can be assigned to MatrixSymbol, but not to Symbol, as
+        the dimensions will not align.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, MatrixSymbol, Matrix
+    >>> from sympy.codegen.ast import Assignment
+    >>> x, y, z = symbols('x, y, z')
+    >>> Assignment(x, y)
+    Assignment(x, y)
+    >>> Assignment(x, 0)
+    Assignment(x, 0)
+    >>> A = MatrixSymbol('A', 1, 3)
+    >>> mat = Matrix([x, y, z]).T
+    >>> Assignment(A, mat)
+    Assignment(A, Matrix([[x, y, z]]))
+    >>> Assignment(A[0, 1], x)
+    Assignment(A[0, 1], x)
+    """
+
+    rel_op = ':='
+    __slots__ = []
+
+    def __new__(cls, lhs, rhs=0, **assumptions):
+        from sympy.matrices.expressions.matexpr import (
+            MatrixElement, MatrixSymbol)
+        from sympy.tensor.indexed import Indexed
+        lhs = _sympify(lhs)
+        rhs = _sympify(rhs)
+        # Tuple of things that can be on the lhs of an assignment
+        assignable = (Symbol, MatrixSymbol, MatrixElement, Indexed)
+        if not isinstance(lhs, assignable):
+            raise TypeError("Cannot assign to lhs of type %s." % type(lhs))
+        # Indexed types implement shape, but don't define it until later. This
+        # causes issues in assignment validation. For now, matrices are defined
+        # as anything with a shape that is not an Indexed
+        lhs_is_mat = hasattr(lhs, 'shape') and not isinstance(lhs, Indexed)
+        rhs_is_mat = hasattr(rhs, 'shape') and not isinstance(rhs, Indexed)
```

Comment:
`NDimArray` also carries a `shape` property, I'm unsure if that has any implications but I thought I should mention it.
If nothing else renaming the variables `*hs_is_mat` to `*hs_has_shape` is more clear and then treat Indexed as a  special case?